lib/logger/lib/logger/legacy_handler.ex

```diff
@@ -0,0 +1,265 @@
+defmodule Logger.LegacyHandler do
+  @moduledoc false
+
+  alias Logger.Counter
+
+  @internal_keys [:counter]
+
+  @doc false
+  def child_spec(start_options) do
+    %{
+      id: :gen_event,
+      start: {:gen_event, :start_link, [{:local, Logger}, start_options]},
+      modules: :dynamic
+    }
+  end
+
+  def adding_handler(%{config: data} = config) do
+    # TODO: When using counters exclusively then this line can be changed to use
+    # `Map.put/2` instead as we will not need to pass ETS table
+    new_data =
+      Logger.Config.default_config()
+      |> Map.merge(data || %{})
+      |> Map.put_new_lazy(:counter, &Logger.Config.new/0)
+
+    {:ok, %{config | config: new_data}}
+  end
+
+  # TODO: Remove when we will support OTP 22+
+  def changing_config(current, new), do: changing_config(:update, current, new)
+
+  def changing_config(
+        op,
+        %{id: id, module: module, config: old_data, filters: filters},
+        %{config: new_data} = new_config
+      ) do
+    {counter, old_data} = Map.pop(old_data, :counter)
+
+    old_data =
+      case op do
+        :set -> Logger.Config.default_config()
+        :update -> old_data
+      end
+
+    data =
+      old_data
+      |> Map.merge(new_data)
+      |> Map.put(:counter, counter)
+
+    config =
+      %{
+        id: id,
+        module: module,
+        level: :all,
+        filter_default: :log,
+        formatter: {:logger_formatter, %{}},
+        filters: filters
+      }
+      |> Map.merge(new_config)
```

Comment:
Maybe we should do: `Map.merge(old_config, new_config)` where `old_config` is the second argument?